server/src/main/java/io/druid/segment/realtime/plumber/RealtimePlumber.java

```diff
@@ -395,11 +425,35 @@ public void persist(final Committer committer)
   {
     final List<Pair<FireHydrant, Interval>> indexesToPersist = Lists.newArrayList();
     for (Sink sink : sinks.values()) {
+      FireHydrant hydrant = sink.swap();
+      if (hydrant != null) {
+        indexesToPersist.add(Pair.of(hydrant, sink.getInterval()));
+      }
+    }
+    persist(committer, indexesToPersist);
+  }
+
+  public void persistOldest(final Committer committer)
+  {
+    Sink oldestSink = null;
+    long oldestAccessTime = -1;
+    for (Sink sink : sinks.values()) {
       if (sink.swappable()) {
-        indexesToPersist.add(Pair.of(sink.swap(), sink.getInterval()));
+        long lastAccessTime = sink.getLastAccessTime();
+        if (oldestAccessTime < 0 || lastAccessTime < oldestAccessTime) {
+          oldestAccessTime = lastAccessTime;
+          oldestSink = sink;
+        }
       }
     }
+    FireHydrant hydrant = oldestSink == null ? null : oldestSink.swap();
+    if (hydrant != null) {
+      persist(committer, Arrays.asList(Pair.of(hydrant, oldestSink.getInterval())));
```

Comment:
#3284 deals with this by persisting all sinks when maxRowsInMemory is surpassed.